src/Type/Php/AbsFunctionDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,115 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Type\Constant\ConstantFloatType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\Type;
+use PHPStan\Type\UnionType;
+use function abs;
+use function count;
+use function max;
+
+class AbsFunctionDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'abs';
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope,
+	): ?Type
+	{
+		$args = $functionCall->getArgs();
+
+		if (!isset($args[0])) {
+			return null;
+		}
+
+		$type = $scope->getType($args[0]->value);
+
+		if ($type instanceof UnionType) {
```

Comment:
this if/else construction feels like something which can be simplified with `TypeTraverser::map`

Reply:
I'm not sure this will simplify things for union types:
* there's a case where the loop returns `null` to avoid overriding the return type of `abs()`. We would need to emulate this from the callback passed to `TypeTraverser::map` which cannot return `null`.
* the loop tries to merge ranges when possible, e.g. `int<-15, -5>|int<10, 20>` → `int<5, 20>`. This does not seem to be trivial to do with `TypeTraverser::map`

Reply:
Solved with last refactoring.